internals/overlord/checkstate/manager.go

```diff
@@ -48,42 +87,114 @@ func (m *CheckManager) NotifyCheckFailed(f FailureFunc) {
 
 // PlanChanged handles updates to the plan (server configuration),
 // stopping the previous checks and starting the new ones as required.
-func (m *CheckManager) PlanChanged(p *plan.Plan) {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
+func (m *CheckManager) PlanChanged(newPlan *plan.Plan) {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// Update local reference to plan.
+	m.planLock.Lock() // always acquire locks in same order (state lock, then plan lock)
+	oldPlan := m.plan
+	m.plan = newPlan
+	m.planLock.Unlock()
+
+	if oldPlan == nil {
+		oldPlan = &plan.Plan{}
+	}
+	shouldEnsure := false
+	newOrModified := make(map[string]bool, len(newPlan.Checks))
+
+	// Abort all currently-running checks.
+	for _, change := range m.state.Changes() {
+		switch change.Kind() {
+		case performCheckKind, recoverCheckKind:
```

Comment:
I believe that with the new design, a new change (recover/perform) created for a particular check will only ever now happen after the existing change is ```ready```, which means we will only ever see one change at most that is non-ready in the loop for any given change, which is nice.

Reply:
Yep, that's right, that is nice.